### ISSUE TITLE
Different lengthscales in different domains

### ISSUE THREAD

**Issue by ankushaggarwal** [cornellius-gp/gpytorch#1253]
I am new to GPytorch and am trying to regress GP to some data which has fast varying and slow varying parts. As a result a single lengthscale cannot capture both. I was wondering if it is possible to have a kernel with two different lengthscales depending.

**Comment by dhruvbalwada** [comment]
I am working on a similar problem. I had thought that using the option ard_num_dim would work. But, I am unable to figure out how to initialize or train these parameters. 

Here is some is an example code: 


```
x= np.random.rand(100, 2)
y = np.sin(x[:,0] + 2*x[:,1])

train_x = torch.from_numpy(x).float()
train_y = torch.from_numpy(y).float()

class ExactGPModel(gpytorch.models.ExactGP):

    def __init__(self, train_x, train_y, likelihood):
        super(ExactGPModel, self).__init__(train_x, train_y, likelihood)  
        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = gpytorch.kernels.ScaleKernel(
                            gpytorch.kernels.RBFKernel(ard_num_dim=2,
                                                      has_lengthscale=True))

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)

likelihood = gpytorch.likelihoods.GaussianLikelihood()
model = ExactGPModel(train_x, train_y, likelihood)

hypers = {
    'likelihood.noise_covar.noise': torch.tensor(.01),
    'covar_module.base_kernel.lengthscale': torch.tensor([[1.,0.],[0., 0.5]]
                                                         ,requires_grad=True),
    'covar_module.outputscale': torch.tensor(1.),
}

model.initialize(**hypers)
```

I get an error `RuntimeError: shape '[1, 1]' is invalid for input of size 4`.

**Comment by dhruvbalwada** [comment]
Oops found a bug in older code, edited above code with working version now.

**Comment by jacobrgardner** [comment]
@dhruvbalwada's suggestion to use ARD is on the money assuming you're just talking about different lengthscales in different dimensions. One very small point: `has_lengthscale` is true by default for RBFKernel, so specifying it should be unnecessary.

If your goal is to achieve different lengthscales in different regions of the space, this effect can often best be achieved by using a small neural net feature extractor before your GP.

If you have a mixture of seasonal trends, the references suggested are good. You could also try spectral kernel learning, which we have a few options for, which can help automate that process (although the initialization can be a bit tricky).

**Comment by ankushaggarwal** [comment]
@jacobrgardner You are right. I am looking for different lengthscales in different regions of the space. Is there an example on how to use neural net feature extractor before GP?

**Comment by dhruvbalwada** [comment]
I am curious to know about the neural network feature extractor too. 
One approach I am aware of is to use a moving window and assuming local stationarity (used here - https://royalsocietypublishing.org/doi/10.1098/rspa.2018.0400 )

**Comment by jacobrgardner** [comment]
If you want to try it out, just have your forward method pass data through a NN first:

```python
def forward(self, x):
    x = self.feature_extractor(x)  # any PyTorch NN
   covar_x = self.covar_module(x)
    ...
```

The NN parameters will automatically get gradient updates during training, since it's all just PyTorch code.
